apps/dashboard/test/system/jobs_app_test.rb

```diff
@@ -274,4 +295,33 @@ def script_yml(dir)
       assert_nil YAML.safe_load(File.read("#{script_dir}/1_job_log"))
     end
   end
+
+  test 'quick launch button submits job and returns job data successfully' do
+    Dir.mktmpdir do |dir|
+      setup_project(dir)
+      setup_script(dir)
+      setup_cache(dir)
+
+      find('[href="/projects/1"]').click
+
+      job = OodCore::Job::Info.new(id: 'job-id-123', status: :completed)
+
+      Open3
+        .stubs(:capture3)
+        .with({}, 'sbatch', '-A', 'pas2051', '--export', 'NONE', '--parsable', '-M', 'owens',
+            { stdin_data: "hostname\n" })
+        .returns(['job-id-123', '', exit_success])
+
+      OodCore::Job::Adapters::Slurm
+        .any_instance
+        .stubs(:info)
+        .returns(job)
+
+      click_on 'Launch'
+      assert_selector('.alert-success', text: 'Successfully submited job job-id-123.')
```

Comment:
@johrstrom I keep getting a failure on this test because the `job_id` is not returning and displaying. Any ideas what I'm `stub`ing wrong or not understanding?

Reply:
Looking at this more, I'm not following what happens in the submit of the controller.

The line which has me stumped is
```ruby
  def submit(options)
    adapter = adapter(options[:cluster]).job_adapter
...
```

Because I expect `adapter` to give back a cluster since it is a hash and seems to want a cluster key, but I'm not sure what "gives back a cluster" really means. Clearly it is somehow related to Ood Core since `job_adapter` is called. And then later that adapter is submited (I'm not sure to what this method is from) with a job_script which gives the `job_id`. 

I need to stub this `adapter` somehow to get that `job_id` back though. I'll keep looking at the gems to see if I can get it.

Reply:
this is the stub that you want. We kinda want to stub it at the lowest layer such that everything in between is actually being tested.

https://github.com/OSC/ondemand/blob/e484eeec58784d78eac2b4034300bb4667323aef/apps/dashboard/test/system/jobs_app_test.rb#L233-L240


You're right. This appears to be named incorrectly, should be `cluster` as that's the object it's returning, not an adapter instance.

https://github.com/OSC/ondemand/blob/e484eeec58784d78eac2b4034300bb4667323aef/apps/dashboard/app/models/script.rb#L273-L275